File: src/net/merayen/elastic/util/tap/Tap.java
package net.merayen.elastic.util.tap;

import java.io.Closeable;

import net.merayen.elastic.util.tap.TapSpreader.Func;

public class Tap<T> implements Closeable {
	private final TapSpreader<T> listener;
	Func<T> func;

	Tap(TapSpreader<T> listener) {
		this.listener = listener;
	}

	public void set(Func<T> func) {
		this.func = func;
	}

	public void close() {
		listener.remove(this);
	}
}

File: src/main/kotlin/net/merayen/elastic/util/tap/Tap.kt
package net.merayen.elastic.util.tap

import java.io.Closeable

/**
 * A tap that taps into a distributor.
 */
class Tap<T> internal constructor(private val listener: ObjectDistributor<T>, val func: (item: T) -> Unit) : Closeable {
	/**
	 * Ends this tap, meaning that it will not call the user function anymore.
	 */
	override fun close() {
		listener.remove(this)
	}
}
